src/WebClient.spec.js

```diff
@@ -500,15 +501,95 @@ describe('WebClient', function () {
   });
 
   describe('has rate limit handling', function () {
-    it('should expose retry headers in the response');
+    it('should expose retry headers in the response', function() {
+      const scope = nock('https://slack.com')
+        .post(/api/)
+        .reply(200, { ok: true }, {
+          'retry-after': 1
+        });
+      const client = new WebClient(token);
+      return client.apiCall('method')
+        .then((result) => {
+          assert.deepNestedInclude(result, { 'retryAfter': 1000 });
+          scope.done();
+        });
+    });
+
     // NOTE: see retry policy note below
-    it('should allow rate limit triggered retries to be turned off');
+    it('should allow rate limit triggered retries to be turned off', function() {
+      const scope = nock('https://slack.com')
+        .post(/api/)
+        .reply(429, { ok: false }, {
+          'retry-after': 0
+        })
+        .post(/api/)
+        .reply(200, { ok: true });
+      const client = new WebClient(token, { retryConfig: { retries: 0 } });
+      return client.apiCall('method')
+        .then(() => {
+          assert.fail(0, 1, 'Rate Limit retry not turned off');
+        })
+        .catch((error) => {
+          assert.isFalse(scope.isDone());
+        });
+    });
+
+
 
     describe('when a request fails due to rate-limiting', function () {
       // NOTE: is this retrying configurable with the retry policy? is it subject to the request concurrency?
-      it('should automatically retry the request after the specified timeout');
-      it('should pause the remaining requests in queue');
-      it('should emit a rate_limited event on the client');
+      it('should automatically retry the request after the specified timeout', function() {
+        const scope = nock('https://slack.com')
+          .post(/api/)
+          .reply(429, { ok: true }, {
+            'retry-after': 1
+          })
+          .post(/api/)
+          .reply(200, { ok: true });
+        const client = new WebClient(token, { retryConfig: { retries: 1 } });
+        return client.apiCall('method')
+          .then((resp) => {
+            assert.propertyVal(resp, 'ok', true);;
+          });
+      });
+
+      it('should pause the remaining requests in queue', function() {
+        const scope = nock('https://slack.com')
+          .post(/api/)
+          .reply(429, { ok: true }, {
+            'retry-after': 0
+          })
+          .post(/api/)
+          .reply(429, { ok: true }, {
+            'retry-after': 1
+          })
+          .post(/api/)
+          .reply(429, { ok: true }, {
+            'retry-after': 0
+          })
+          .post(/api/)
+          .reply(200, { ok: true });
+        const client = new WebClient(token, { retryConfig: { retries: 3 } });
+        return client.apiCall('method')
```

Comment:
i don't think this reflects the intention of the test case. the idea is: given a WebClient with a request concurrency of 2 (just to keep things simple), you fire three requests concurrently (without waiting for the promises to resolve). the first request gets rate-limited, for a specified amount of time, let's say 1 second. the second request would have already been triggered (the concurrency allows for 2 requests to be in flight at the same time) and let's say it completes very quickly (~10ms), but the third request should not even start until after the 1 second timeout from the first request has elapsed. does this make sense?

Reply:
i get it, will change to reflect this